lib/cloud-operation/cloud-operation-v2.ts

```diff
@@ -0,0 +1,94 @@
+import { CloudOperationBase } from "./cloud-operation-base";
+import { CloudOperationMessageTypes, CloudCommunicationEvents } from "../constants";
+
+class CloudOperationV2 extends CloudOperationBase implements ICloudOperation {
+	private static readonly WAIT_TO_START_TIMEOUT: number = 10 * 60 * 1000;
+
+	private communicationChannel: ICloudCommunicationChannel;
+	private waitToStartTimeout: NodeJS.Timer;
+	private waitResultPromise: Promise<ICloudOperationResult>;
+
+	constructor(protected id: string,
+		protected serverResponse: IServerResponse,
+		private $nsCloudCommunicationChannelFactory: ICloudCommunicationChannelFactory) {
+		super(id, serverResponse);
+
+		this.communicationChannel = this.$nsCloudCommunicationChannelFactory.create(serverResponse.communicationChannel, this.id);
+	}
+
+	public async sendMessage<T>(message: ICloudOperationMessage<T>): Promise<void> {
+		this.isInitialized();
+		await this.communicationChannel.sendMessage(message);
+	}
+
+	public async cleanup(exitCode?: number): Promise<void> {
+		if (this.waitToStartTimeout) {
+			clearTimeout(this.waitToStartTimeout);
+			this.waitToStartTimeout = null;
+		}
+
+		await this.communicationChannel.close(exitCode);
```

Comment:
Shouldn't we remove the attached handlers as well?

Reply:
1. Yes, we should.
2. I removed the invocation of the cleanup method from within the cloud operation. Now it should be called by the user of the cloud operation instance. Do you think this is ok?